bin/create_mirror.py

```diff
@@ -0,0 +1,190 @@
+
+
+class ACLMirrorer:
+    def __init__(self, args) -> None:
+        self.hash_mismatches = []
+        self.not_downloadable = []
+        self.args = args
+        self.source = args["--source"]
+        self.to = args["--to"]
+        self.is_dry_run = args["--dry-run"]
+        self.only_papers = args["--only-papers"]
+
+    def download_file(self, fname: str, hash: str, type: str) -> None:
+        if type == "pdf":
+            remote_url = self.source + "/" + fname
+        elif type == "attachment":
+            remote_url = self.source + "/attachments/" + fname
+        else:
+            log.error("unrecognized type: " + type)
+            exit(1)
+        local_target = ""
+        # mkstemp opens the file but we only need the file name, so close it again.
+        tmpfd, tmp_target = tempfile.mkstemp(prefix="aclmirrorer_", suffix=".pdf")
+        os.close(tmpfd)
+        match = NEW_ID_RE.match(fname)
+        if match:
+            local_target = os.path.join(self.to, type, match.groups()[1], fname)
+        else:
+            match = OLD_ID_RE.match(fname)
+            if match:
+                local_target = os.path.join(
+                    self.to,
+                    type,
+                    match.groups()[0],
+                    match.groups()[0] + match.groups()[1],
+                    fname,
+                )
+            else:
+                log.error("unrecognized format for " + fname)
+                exit(1)
+
+        if os.path.exists(local_target):
+            existing_hash = anthology_utils.compute_hash_from_file(local_target)
+            if existing_hash == hash:
+                log.debug(
+                    "File {} already up to date, not downloading again".format(
+                        local_target
+                    )
+                )
+                return
+            else:
+                log.debug("File {} changed, redownloading ...".format(local_target))
+        else:
+            log.debug("Downloading {} from {} ...".format(local_target, remote_url))
+
+        if self.is_dry_run:
+            return
+
+        local_path = os.path.dirname(local_target)
+        os.makedirs(local_path, exist_ok=True)
+        try:
+            urlretrieve(remote_url, tmp_target)
+        except:
+            log.error("could not download " + remote_url)
+            self.not_downloadable.append(remote_url)
+            os.remove(tmp_target)
+            return
+        new_hash = anthology_utils.compute_hash_from_file(tmp_target)
+        if new_hash == hash:
+            # all good, store downloaded file in the proper place
+            shutil.move(tmp_target, local_target)
+        else:
+            log.error(
+                "Hash mismatch for file {}, downloaded from {}. was {} should be {}".format(
+                    local_target, remote_url, new_hash, hash
+                )
+            )
+            self.hash_mismatches.append(remote_url)
+        return
+
```

Comment:
Any chance this could be refactored to use `anthology.util.retrieve_url`?

Reply:
Is that comment meant for the downloading? seems to be at an incorrect line.

I used `urlretrieve` because 1) it is built-in in python 2) `retrieve_url` has functionality we don't need (cookies) and does not have functionality we need (https) and has overhead over `urlretrieve`.